ollama: 0.1.17 -> 0.1.24

abysssol opened the pull request:
## Description of changes

Update ollama from version `0.1.17` to version `0.1.24`.

I'm new to nixpkgs, so review carefully to ensure that I didn't accidentally do something wrong.

Resolves #280030

### A collection of ollama's release notes

The new version appears backward compatible, but I'm not entirely sure.
_Click to expand details._
<details><summary>

### _ollama release [v0.1.18](https://github.com/ollama/ollama/releases/tag/v0.1.18)_

</summary>

## What's Changed

- Added `/? shortcuts` help command to `ollama run` to list keyboard shortcuts
- Improved performance when sending follow up messages in `ollama run` or via the API.
- Fixed issues where certain 7B models would error on GPUs with 4GB of memory or less
- Fixed issue where Llava model prompts couldn't start with a file path
- Fixed issue where model would not be correctly reloaded if options or parameters changed between requests
- Ollama will now automatically pull new models when running older ggml format models. If using custom ggml format models in a Modelfile, please [import GGUF](https://github.com/jmorganca/ollama/blob/main/docs/import.md) models instead.

</details><details><summary>

### _ollama release [v0.1.19](https://github.com/ollama/ollama/releases/tag/v0.1.19)_

</summary>

## What's Changed

- Fixed "out of memory" errors when running models such as llama2, mixtral or llama2:13b with limited GPU memory
- Fixed CUDA errors when running on older GPUs that aren't yet supported
- Increasing context size with num_ctx will now work (up to a model's supported context window).
- Larger models such as mixtral can now be run on Macs with less memory
- Fixed an issue where pressing up or down arrow keys would cause the wrong prompt to show in ollama run
- Fixed performance issues on Intel Macs
- Fixed an error that would occur with old Nvidia GPUs
- OLLAMA_ORIGINS now supports browser extension URLs
- Ollama will now offload more processing to the GPU where possible

</details><details><summary>

### _ollama release [v0.1.20](https://github.com/ollama/ollama/releases/tag/v0.1.20)_

</summary>

## What's Changed

- Fixed additional cases where Ollama would fail with `out of memory` CUDA errors
- Multi-GPU machines will now correctly allocate memory across all GPUs
- Fixed issue where Nvidia GPUs would not be detected by Ollama

</details><details><summary>

### _ollama release [v0.1.21](https://github.com/ollama/ollama/releases/tag/v0.1.21)_

</summary>

## What's Changed

- Models can now be saved and loaded with `/save <model>` and `/load <model>` when using `ollama run`. This will save or load conversations and any model changes with `/set parameter`, `/set system` and more as a new model with the provided name.
- Support for a much broader set of CPUs, including CPUs without AVX instruction set support
- If a GPU detection error is hit when attempting to run a model, Ollama will fallback to CPU
- Fixed issue where generating responses with the same prompt would hang after around 20 requests
- New MESSAGE Modelfile command to set the conversation history when building a model
- Ollama will now use AVX2 for faster performance if available
- Improved detection of Nvidia GPUs, especially in WSL
- Fixed issue where models with LoRA layers may not load
- Fixed incorrect error that would occur when retrying network connections in `ollama pull` and `ollama push`
- Fixed issue where `/show parameter` would round decimal numbers
- Fixed issue where upon hitting the context window limit, requests would hang

</details><details><summary>

### _ollama release [v0.1.22](https://github.com/ollama/ollama/releases/tag/v0.1.22)_

</summary>

## What's Changed

- Fixed issue with Nvidia GPU detection that would cause Ollama to error instead of falling back to CPU
- Fixed issue where AMD integrated GPUs caused an error

</details><details><summary>

### _ollama release [v0.1.23](https://github.com/ollama/ollama/releases/tag/v0.1.23)_

</summary>

## What's Changed

- New keep_alive API parameter to control how long models stay loaded
- Image paths can now be provided to `ollama run` when running multimodal models
- Fixed issue where downloading models via `ollama pull` would slow down to 99%
- Fixed error when running Ollama with Nvidia GPUs and CPUs without AVX instructions
- Support for additional Nvidia GPUs (compute capability 5)
- Fixed issue where system prompt would be repeated in subsequent messages
- `ollama serve` will now print prompt when OLLAMA_DEBUG=1 is set
- Fixed issue where exceeding context size would cause erroneous responses in `ollama run` and the /api/chat API
- `ollama run` will now allow sending messages without images to multimodal models

</details><details><summary>

### _ollama release [v0.1.24](https://github.com/ollama/ollama/releases/tag/v0.1.24)_

</summary>

## What's Changed

- Fixed issue where requests to /api/chat would hang when providing empty user messages repeatedly
- Fixed issue on macOS where Ollama would return a missing library error after being open for a long period of time

</details>

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - Personally tested that rocm works
  - Others have tested that cuda works
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

abysssol committed:
maintainers: add abysssol

abysssol committed:
ollama: 0.1.17 -> 0.1.24

Remove outdated patches and substitution of llama.cpp server.

Patch source build script to remove git.
Ollama's `go generate` script assumes the source is an
initialized git repository, which it is not in nix.
So, remove usage of `git` and replace it with `patch`.
Also replace call to `g++` with `$CXX`.

Run `go generate` to build dependencies using
ollama's custom build script.

abysssol committed:
ollama: add support for gpu acceleration

drupol commented:
Do you think we can fix the builds for Darwin arch?

Samasaur1 commented on pkgs/tools/misc/ollama/default.nix:
```suggestion
    nativeBuildInputs = [ cmake ]
      ++ (lib.optional gpuIsEnabled makeWrapper)
      ++ (lib.optional stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
        Accelerate
        Metal
      ]));
```
This will provide Apple's Metal and Accelerate libraries, which are the first error encountered when attempting to build on Darwin.

Samasaur1 replied:
whoops, I definitely screwed this up in the GitHub UI, but hopefully it is clear where this should go

Samasaur1 replied:
You'll also have to add `darwin` as an argument

drupol replied:
Please use `lib.optionals` instead.

Samasaur1 commented:
I tried to build on aarch64-darwin, and got errors trying to access the Accelerate and Metal libraries. I was able to fix those errors with the change I suggested above, but now I see this at the end of the log:
```
+ mkdir -p ../llama.cpp/build/darwin/arm64/metal/lib/
+ g++ -fPIC -g -shared -o ../llama.cpp/build/darwin/arm64/metal/lib/libext_server.dylib -arch arm64 -Wl,-force_load ../llama.cpp/build/darwin/arm64/metal/examples/server/libext_server.a ../llama.cpp/build/darwin/arm>
./gen_common.sh: line 57: g++: command not found
llm/generate/generate_darwin.go:3: running "sh": exit status 127
/nix/store/zf8hff9bnlfix7l7gjsnv3jc22lxzzxx-stdenv-darwin/setup: line 140: pop_var_context: head of shell_variables not a function context
```

drupol commented:
> ./gen_common.sh: line 57: g++: command not found

`g++` is not available on Darwin ?

n8henrie commented:
> g++ is not available on Darwin ?

It is kind of, it's just clang under the covers.

```console
$ sw_vers
ProductName:            macOS
ProductVersion:         14.2.1
BuildVersion:           23C71
$ arch
arm64
$ file /usr/bin/g++
/usr/bin/g++: Mach-O universal binary with 2 architectures: [x86_64:\012- Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>] [\012- arm64e (caps: 0x2):\012- Mach-O 64-bit arm64e (caps: PAC00) executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>]
$ /usr/bin/g++ --version
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: arm64-apple-darwin23.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

viraptor commented:
Was there a specific reason to not use the already-packaged llama-cpp? It looks like we're doing a bit of extra work here now, but I'm not sure what do we gain that way.

drupol commented:
I guess that's because now llama-cpp is a git submodules fixed to a
specific commit?

On Thu, Feb 15, 2024, 23:45 Stanisław Pitucha ***@***.***>
wrote:

> Was there a specific reason to not use the already-packaged llama-cpp? It
> looks like we're doing a bit of extra work here now, but I'm not sure what
> do we gain that way.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/289108#issuecomment-1947459975>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB5RCTCLSXMFYTDI6RVWPDYT2FXPAVCNFSM6AAAAABDKWELMOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNBXGQ2TSOJXGU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

viraptor commented:
I hope we can reuse the llama-cpp to unvendor it. It would make the build simpler. Will report back if that fails.

drupol commented:
I agree it would be definitely better, but we might have incompatibilities with this project.

elohmeier commented:
I think we should keep the llama-cpp part in a separate package to not increase the packaging effort. llama-cpp is useful even without ollama, that's why I put it into a separate package in the first place. See #257760

viraptor commented:
@elohmeier Neither option would remove llama-cpp as a package. The question is only whether to keep a different vendored one in ollama.

reckenrode commented:
If the packaged version doesn’t work, it might be possible to override its src to use the commit pended by ollama. It’ll depend on how ollama integrates it with its own build process.

viraptor commented:
Things seem to work with the packaged one, but... slower than the old version. I can't tell if it's a real regression or something I've done. This is a completely different change, so I've opened https://github.com/NixOS/nixpkgs/pull/289165 instead.

abysssol commented:
I added the apple frameworks to `buidInputs` since they seem like libraries. I can change it back if there's some particular need for them to be in `nativeBuildInputs`.
I also squashed the commit history since @drupol suggested it (on matrix).

abysssol commented:
> Was there a specific reason to not use the already-packaged llama-cpp? It looks like we're doing a bit of extra work here now, but I'm not sure what do we gain that way.

I used the vendored submodule of llama-cpp because:
1. Ollama's custom build script does patching of llama-cpp's source, so I don't think the existing nix package would work.
2. It was simpler than trying to splice together a separate `fetchFromGithub` source and applying necessary patches.

abysssol commented:
@Samasaur1 by the way I did patch ollama's build script to use `$CXX` instead of `g++`.

Samasaur1 commented:
> I added the apple frameworks to `buidInputs` since they seem like libraries. I can change it back if there's some particular need for them to be in `nativeBuildInputs`. I also squashed the commit history since @drupol suggested it (on matrix).

I tried building again on aaeaec2ed7f02e49172624c245933397eaec2d6d (with `nix-build -A ollama`) and I'm back to seeing errors for Metal and Accelerate being missing

abysssol commented:
@Samasaur1 ok sorry I changed it back

Samasaur1 commented:
Apparently MetalKit is a separate framework — I'm adding that and trying again

Samasaur1 commented:
Okay, with Accelerate, Metal, MetalKit, and MetalPerformanceShaders in `nativeBuildInputs`, then I am able to build that specific dependency. However, later on in the build script I *still* get errors about Accelerate and Metal being missing (probably in another dependency?)

Samasaur1 commented:
Here's my current log: https://pastebin.com/1Amcn8Ch

Samasaur1 commented:
Okay, I got it! These frameworks do need to be in `buildInputs` instead of `nativeBuildInputs`, but they need to be in the `buildInputs` block of `ollama`, and they were in the `buildInputs` block of `gpuBuildLibs`

Samasaur1 commented on pkgs/tools/misc/ollama/default.nix:
```diff
@@ -1,50 +1,167 @@
+    };
+    vendorHash = "sha256-wXRbfnkbeXPTOalm7SFLvHQ9j46S/yLNbFy+OWNSamQ=";
+
+    nativeBuildInputs =
```
```suggestion
    nativeBuildInputs =
      [ cmake ]
      ++ (lib.optional gpuIsEnabled makeWrapper);

    buildInputs = lib.optionals stdenv.isDarwin [
      Accelerate
      Metal
      MetalKit
      MetalPerformanceShaders
    ];
```

Samasaur1 reviewed (changes requested):
I have probably screwed up the GitHub UI again but I think this is the only change needed to build on Darwin

Samasaur1 commented:
Latest commit builds for me!

abysssol commented:
@Samasaur1 sorry, it seems `gpuBuildLibs` wasn't updating `ollama`'s attributes because it's optional on `gpuIsEnabled`, which I forgot to update to include building on darwin. It was my oversight.

abysssol commented:
@Samasaur1 can you test whether ollama actually finds and uses the gpu for you?

Samasaur1 commented:
> @Samasaur1 can you test whether ollama actually finds and uses the gpu for you? It should log whether it can from:
> 
> ```shell
> ollama serve
> ```

@abysssol this is what I see
```
$ ./result/bin/ollama serve
Couldn't find '/Users/sam/.ollama/id_ed25519'. Generating new private key.
Your new public key is:

ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJ+Bumd2WyaExUzw8sU9UXP4S3AEJbDBcvOP2Gpo6mzG

time=2024-02-15T18:10:25.781-08:00 level=INFO source=images.go:863 msg="total blobs: 0"
time=2024-02-15T18:10:25.781-08:00 level=INFO source=images.go:870 msg="total unused blobs removed: 0"
time=2024-02-15T18:10:25.782-08:00 level=INFO source=routes.go:999 msg="Listening on 127.0.0.1:11434 (version 0.1.24)"
time=2024-02-15T18:10:25.782-08:00 level=INFO source=payload_common.go:106 msg="Extracting dynamic libraries..."
time=2024-02-15T18:10:25.805-08:00 level=INFO source=payload_common.go:145 msg="Dynamic LLM libraries [metal]"
[GIN] 2024/02/15 - 18:10:56 | 200 |     165.167µs |       127.0.0.1 | GET      "/"
[GIN] 2024/02/15 - 18:10:56 | 404 |         6.5µs |       127.0.0.1 | GET      "/favicon.ico"
```

abysssol commented:
@Samasaur1 could you try running this and uploading `ollama.log`?
``` sh
export OLLAMA_DEBUG=1; ollama serve &> ollama.log & ollama run dolphin-phi
```
It'll probably take a minute or two to download the model.

viraptor commented:
@abysssol
> Ollama's custom build script does patching of llama-cpp's source, so I don't think the existing nix package would work.

This is easy to do. Done in the other PR.

Samasaur1 commented:
@abysssol here: [ollama.log](https://github.com/NixOS/nixpkgs/files/14304628/ollama.log)

Samasaur1 commented:
I'm fairly certain that it is using the GPU, though, because I watched the GPU usage go up in Activity Monitor when I started asking questions:
<img width="346" alt="Screenshot 2024-02-15 at 6 34 57 PM" src="https://github.com/NixOS/nixpkgs/assets/30577766/046e193f-e7ad-4129-b70b-3e17d0b2d1b1">

abysssol commented:
@viraptor are you aware of how ollama patches llama-cpp with `include (../../../ext_server/CMakeLists.txt)`, which splices ollama's files (from outside llama-cpp's directory) into llama-ccp's build process? This would have to be accounted for, and the path to this file in the nix store may have to be patched in. Maybe it would be best to add an input to the llama-cpp module that appends custom patches before building the server and libraries?

abysssol commented:
@Samasaur1 thank you. yes, it seems to be finding the gpu.
as per the logs:
```
ggml_backend_metal_buffer_from_ptr: allocated buffer, size =  1456.20 MiB, ( 1456.27 / 21845.34)
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:        CPU buffer size =    70.31 MiB
llm_load_tensors:      Metal buffer size =  1456.20 MiB
```

abysssol commented:
@Samasaur1 are you sure `darwin.apple_sdk.frameworks.MetalPerformanceShaders` is a real and required package? ofborg found that it doesn't exist, and when i tried installing the framework files, the package attribute didn't exist for x86_64-linux, while the `darwin.apple_sdk.frameworks.Accelerate` package existed and built.

Samasaur1 commented:
> @Samasaur1 are you sure `darwin.apple_sdk.frameworks.MetalPerformanceShaders` is a real and required package? ofborg found that it doesn't exist, and when i tried installing the framework files, the package attribute didn't exist for x86_64-linux, while the `darwin.apple_sdk.frameworks.Accelerate` package existed and built.

@abysssol it was required for me when the dependencies were in `nativeBuildInputs`. I'll try removing them and rebuilding

Samasaur1 commented:
Yeah, it seems to be real and necessary. Maybe just on Apple Silicon though? You could make it a conditional dependency that's only present if the platform is `aarch64-darwin`

abysssol commented:
> You could make it a conditional dependency that's only present if the platform is `aarch64-darwin`

@Samasaur1 I thought I already was by using `optional`/`optionalAttrs`? Could you give me an example of doing it right?

Samasaur1 commented:
> > You could make it a conditional dependency that's only present if the platform is `aarch64-darwin`
> 
> @Samasaur1 I thought I already was by using `optional`/`optionalAttrs`? Could you give me an example of doing it right?

@abysssol you are correctly doing optional dependencies if the platform is Darwin, but I mean doing it specifically if the platform is Darwin on aarch64 (i.e., Apple Silicon). Like this:
```nix
  appleFrameworks = darwin.apple_sdk.frameworks;
  gpuBuildLibs = {
    buildInputs = (lib.optionals rocmIsEnabled [
      rocmPackages.clr
      rocmPackages.hipblas
      rocmPackages.rocblas
      rocmPackages.rocsolver
      rocmPackages.rocsparse
      libdrm
    ])
    ++ (lib.optional cudaIsEnabled
      cudaPackages.cuda_cudart
    )
    ++ (lib.optionals stdenv.isDarwin [
      appleFrameworks.Accelerate
      appleFrameworks.Metal
    ])
    ++ (lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
      appleFrameworks.MetalKit
      appleFrameworks.MetalPerformanceShaders
    ]);
  };
```

Samasaur1 commented:
actually I just checked in `nix repl` and given that `MetalKit` does *exist* on x86_64-darwin, I'm pretty sure it'd be required there too. so:
```nix
  appleFrameworks = darwin.apple_sdk.frameworks;
  gpuBuildLibs = {
    buildInputs = (lib.optionals rocmIsEnabled [
      rocmPackages.clr
      rocmPackages.hipblas
      rocmPackages.rocblas
      rocmPackages.rocsolver
      rocmPackages.rocsparse
      libdrm
    ])
    ++ (lib.optional cudaIsEnabled
      cudaPackages.cuda_cudart
    )
    ++ (lib.optionals stdenv.isDarwin [
      appleFrameworks.Accelerate
      appleFrameworks.Metal
      appleFrameworks.MetalKit
    ])
    ++ (lib.optional (stdenv.isDarwin && stdenv.isAarch64)
      appleFrameworks.MetalPerformanceShaders
    );
  };
```

viraptor commented on pkgs/tools/misc/ollama/patch/remove-git.patch:
```diff
@@ -0,0 +1,71 @@
+-    git submodule update --force ${LLAMACPP_DIR}
+-
+-}
++git_module_setup() { :; }
```
Replacing the whole thing is brittle and will cause conflicts if upstream changes anything. How about setting `OLLAMA_SKIP_PATCHING` instead for the build, so the function doesn't run?

viraptor commented on pkgs/tools/misc/ollama/patch/remove-git.patch:
```diff
@@ -0,0 +1,71 @@
+-    # Avoid duplicate main symbols when we link into the cgo binary
+-    sed -e 's/int main(/int __main(/g' <${LLAMACPP_DIR}/examples/server/server.cpp >${LLAMACPP_DIR}/examples/server/server.cpp.tmp &&
+-        mv ${LLAMACPP_DIR}/examples/server/server.cpp.tmp ${LLAMACPP_DIR}/examples/server/server.cpp
++    patch -i '@cmakeIncludePatch@' "${LLAMACPP_DIR}/examples/server/CMakeLists.txt"
```
Similar - we'll need to keep resolving conflicts here. How about:
```
    patch -i '@cmakeIncludePatch@' "${LLAMACPP_DIR}/examples/server/CMakeLists.txt"
    return

    # rest of the function unchanged
```
at the top?

viraptor commented on pkgs/tools/misc/ollama/patch/remove-git.patch:
```diff
@@ -0,0 +1,71 @@
+-        done
+-    fi
+-}
++cleanup() { :; }
```
Same:
```
return
```
at the top should do the trick

viraptor commented:
I raised the issue upstream and summarised why the build as it is is a bad idea: https://github.com/ollama/ollama/issues/2534

I really wish we didn't have to do the duplication between those packages. This setup will get out of sync and cause inconsistencies sometime in the future and it would be great to avoid that.

abysssol commented:
> Replacing the whole thing ... will cause conflicts if upstream changes anything.

@viraptor That's the point. If upstream changes the build script enough to break the patches, then I should create new patches after carefully inspecting the new version of the build script for correct, intended behavior, instead of letting potentially outdated patches apply.

For example, notice how the build script says:
`ollama/llm/generate/gen_common.sh` -> "apply temporary patches until fix is upstream".
This implies that future versions of the build script will no longer apply certain patches, though may apply others instead.

As an avid rust enjoyer, I like the compiler/computer telling me when something that I assumed would be true has changed, and that I should fix it (such as a major update to a library).

I value correctness above all else (though efficiency is a close second).

viraptor commented:
We still need to check that the package works on every update. If something changed that will break it - you'll know at that point. But if there's only churn, things will continue working and you'll avoid needing to refresh the patches. By very specific big patches, you're signing up current maintainers to deal the churn even when not needed.

On the other hand, you're not preventing unexpected changes outside of your patch file from breaking the build - you'd have to validate the whole file for that.

If you want more correctness, a better solution for that would be to add automatic tests (even querying the API without any model).

abysssol commented:
Ok fine, you make some good points. I guess building software with an unholy agglomeration of shell, c/c++, cmake, and go doesn't lend itself well to machine checked correctness. I'll change the patches.

abysssol commented:
> If you want more correctness, a better solution for that would be to add automatic tests (even querying the API without any model).

This sounds interesting, but I don't know how. Could you elaborate?

viraptor commented:
There's more than I can fit here in a reasonable time :)
Have a look at the descriptions of testers https://nixos.org/manual/nixpkgs/stable/#tester-nixosTest
And a few examples in `nixos/tests/...` - those spawn a new vm with the service installed (ollama already has a service module available), so you can interact with the service in any way you want. That will get run automatically when the package is changed - you can catch any unexpected behaviour there.

TanvirOnGH commented:
v0.1.25 is out: https://github.com/ollama/ollama/releases/tag/v0.1.25

drupol commented on pkgs/tools/misc/ollama/default.nix:
```diff
@@ -1,50 +1,177 @@
+    inherit pname version src;
+    vendorHash = "sha256-wXRbfnkbeXPTOalm7SFLvHQ9j46S/yLNbFy+OWNSamQ=";
+
+    nativeBuildInputs = [ cmake ]
+      ++ (lib.optional enableLinuxGpu makeWrapper);
```
```suggestion
    nativeBuildInputs = [ 
      cmake
    ] ++ lib.optionals enableLinuxGpu [
      makeWrapper
    ];
```

drupol commented on pkgs/tools/misc/ollama/default.nix:
```diff
@@ -1,50 +1,177 @@
+  };
+
+  linuxGpuLibs = {
+    buildInputs = (lib.optionals rocmIsEnabled [
+      rocmPackages.clr
+      rocmPackages.hipblas
+      rocmPackages.rocblas
+      rocmPackages.rocsolver
+      rocmPackages.rocsparse
+      libdrm
+    ])
+    ++ (lib.optional cudaIsEnabled
+      cudaPackages.cuda_cudart
+    );
```
```suggestion
    buildInputs = lib.optionals rocmIsEnabled [
      rocmPackages.clr
      rocmPackages.hipblas
      rocmPackages.rocblas
      rocmPackages.rocsolver
      rocmPackages.rocsparse
      libdrm
    ] ++ lib.optionals cudaIsEnabled [
      cudaPackages.cuda_cudart
    ];
```

drupol commented on pkgs/tools/misc/ollama/default.nix:
```diff
@@ -1,50 +1,177 @@
+    ];
+  };
+
+  runtimeLibs = (lib.optional rocmIsEnabled rocmPackages.rocm-smi)
+    ++ (lib.optional cudaIsEnabled linuxPackages.nvidia_x11);
```
```suggestion
  runtimeLibs = lib.optionals rocmIsEnabled [
    rocmPackages.rocm-smi
  ] ++ lib.optionals cudaIsEnabled [
    linuxPackages.nvidia_x11
  ];
```

mschwaig commented:
I tried building different configurations of llama.cpp not too long ago and annoyingly, at least for some of the hardware acceleration options that I tried, it falls back to building the slow version if it does not detect what's required for the specific configured method of hardware acceleration. I would expect this to be a real pain to get to work in the build sandbox.

mschwaig commented:
Regarding ollama, what I remember reading about how they added ROCm support, they build both a CUDA and a ROCm version of llama.cpp, and link to those two binaries in different ways (also renaming some symbols for that purpose), and then at runtime call into both of them to find the working one.

IMO we would not need any of that logic in nixpkgs. I'm assuming we would want to just give hem the one llama.cpp version we want them to run with. That should make more sense for us in almost any case.

I have not actually read up on all of this again to verify this info, but the PR that originally implemented ROCm support and how they build llama.cpp now should be this one: https://github.com/ollama/ollama/pull/1146

Since llama.cpp recently got a number of different backends I would imagine that approach might become an issue for them as well.

abysssol commented:
@mschwaig do you have any idea/suggestion for what that means for this particular pull request? I'm not too familiar with nix, ollama, or llama.cpp, so any implications of what you said are somewhat lost on me.

Are you suggesting that it would be possible to skip building llama.cpp using ollama's build script, and instead expose libraries built with nix? Or something else? Try to get upstream ollama to not use the dynamic library selection anymore?

mschwaig commented:
My approach would be to fork ollama to skip building llama.cpp and bypass that runtime logic for selecting the correct version of llama.cpp.

I would implement this by adding an environment variable as an input to the ollama build which can point to a prebuilt llama.cpp.
If the environment variable is unset the build proceeds as normal.
If the environment variable is set, it enacts the changes described above at build time, which should be enough to make it build in nixpkgs.

Now for actually getting that merged in nixpkgs I would create a patch from that fork. If there were any changes to llama.cpp which ollama needs, I would make those into patches for llama.cpp which I would apply with an override.

Now at the same time as turning that fork into patches, I would also create a PR to try to upstream those changes to ollama. I suspect that they would be useful for the ollama project as well, since if they ever decide to switch to offering different builds for different accelerators they can easily do that with such an environment variable. I think that this would be much less of a headache for them to maintain.
Maybe with llama.cp supporting so many different backends, they change their minds at some point and start providing different builds for each supported backend. They can easily do that with by building binaries that link different versions of ollama.

If ollama accepts the changes from the PR, then the patch is not/no longer needed.
The thing that I am least confident about in this explanation is the environment variable aspect, because I don't know how the Go build system best accepts inputs like that. I have done very similar things with Rust, Gradle and Python builds, but never Go.

PS: @abysssol I think its great that you created a flake for this, and I know you were encouraged to contribute that here, and I think we are lucky to have you. If you feel overwhelmed by the process or you have any issues we can help you with, continue to let us know. You do not owe any of us any work, and I hope you keep having fun working on this PR. This is really not an easy first contribution at all so it's completely understandable if you don't get what I'm saying and I'm happy to clarify.

abysssol commented:
@mschwaig Thank you, I think I understand what you are suggesting, at least at a high level. Though since I don't know that much about ollama's implementation, I currently don't have a good idea of what would need to change to get it working right. I'll probably have to start reading ollama's source and get a decent understanding of how it actually works, and what it actually requires from llama.cpp, which could potentially take weeks.

My current thought is that it would be good to get this pull request merged soon, so it can get to nixpkgs-unstable quickly and actually be usable/useful (it's working correctly now as far as I can tell). After that, and maybe backporting to latest stable (what do you think?), I could try to implement your described changes to ollama's build system in a separate pull request.

I feel like the longest we should wait before merging this would be until ollama's next release, which fixes [an issue](https://github.com/ollama/ollama/issues/2502) where 0.1.25 won't find/use amd gpus on nix due to a missing driver file, `/sys/module/amdgpu/version`.
It is probably worth waiting for 0.1.26, though, because 0.1.24 introduced a performance problem where loading a model with `ollama run <model>` will take far longer than it should, 10 to 60+ seconds, before you can start prompting.

Any thoughts?

Oh also, `ofborg` has been `pending` for a day or two now, does anyone know why it's not finishing it's tests, and how to get it to do so?

drupol commented:
@ofborg build ollama

drupol commented:
I agree with your approach @abysssol .

Regarding ofborg, feel free to amend an empty commit to trigger it when it's stuck.

abysssol commented:
@drupol Ah, I see. So is the `ofborg build ollama` command you did preferred for this sort of situation?

drupol commented:
It's not really preferred, it's just usually faster.

drupol commented:
@viraptor @Samasaur1 @dit7ya @elohmeier Are you all OK to merge this as is ?

drupol commented:
I'm planning to merge this as is at the end of the work day (3:00 PM GMT / UTC) if no objection.

drupol commented:
@abysssol Welcome to the Nix contributor family ! Thank you very much for this contribution, I'm so glad you made your first contribution, and not a small one!

To be honest, I hope it's not the last !!!

abysssol commented:
@drupol Thank you; I don't expect it to be the last. At the very least, I plan to keep ollama up to date.
I'm also going to try looking into compiling ollama without it's custom build script, using libraries built with nix.

I want to backport a newer ollama version to the latest stable as well, but I'm don't think that 0.1.24 is a good candidate given it's startup performance problem, so I'll probably wait for version 0.1.26 to release. It might actually be a good idea to update to 0.1.25 (master and stable), since I expect most users won't even enable gpu acceleration, or know that it's possible, so won't be impacted by that version's inability to use amd gpus.
Could it be possible to automatically enable gpu acceleration if a graphics card is detected, like if the kernel has gpu driver modules enabled? Or is that information not available within nixpkgs?

mschwaig commented:
> Could it be possible to automatically enable gpu acceleration if a graphics card is detected, like if the kernel has gpu driver modules enabled? Or is that information not available within nixpkgs?

Doing that kind of thing from inside a Nix package goes against how Nix wants to do things, because you want to abstract away the build host and get the same output every time. So in nixpkgs it makes sense to have an explicit switch for these kinds of things and decide deliberately about enabling or disabling it.

> I'm also going to try looking into compiling ollama without it's custom build script, using libraries built with nix.

Take a look at https://github.com/NixOS/nixpkgs/pull/289165 for that if you have not seen it.

happysalada commented:
@abysssol hey, thank you for this PR !

Actually, I've been working on llama-cpp to include all the shared libs out of the box for ollama.
Potentially, we don't need to fetch-submodule and build llama-cpp, we could just ln -s the output from llama-cpp into the right directory. There might be a bit of a drift in terms of versions, but over the commits, things should largely remain compatible.
Of course since you are a maintainer, it's your decision. I thought it might simplify the process.

abysssol commented:
@happysalada yes, that sounds exactly like what I was imagining/planning, the only question being whether llama-cpp is being built in a way that is compatible with what ollama expects. Have you looked at ollama's build script? Besides a couple patches, ollama also modifies `${llama-cpp}/examples/server/CMakeLists.txt` to include their own`${ollama}/llm/ext_server/CMakeLists.txt`, which adds `${ollama}/llm/ext_server/ext_server.cpp` as a library. And I'm not sure what other customization ollama might have made, since I haven't yet had the time to thoroughly read through ollama's build script.

While this sort of integration is my end goal, I'm not convinced that it'll be simple to get working.

happysalada commented:
Oh i had only seen the dynamic dependencies on *.so files.
Understood regarding quirks.
If there is something you expect more out of llama.cpp, feel free to ping me. At least now, all the .so files are in the derivation.

SomeoneSerge commented on pkgs/tools/misc/ollama/default.nix:
```diff
@@ -1,50 +1,182 @@
+
+  runtimeLibs = lib.optionals rocmIsEnabled [
+    rocmPackages.rocm-smi
+  ] ++ lib.optionals cudaIsEnabled [
+    linuxPackages.nvidia_x11
+  ];
```
Hi! Could you confirm if you used `linuxPackages.nvidia_x11` to link `libcuda.so`? If so, you probably want to remove this and add `autoAddDriverRunpath` to `nativeBuildInputs`: `${linuxPackages.nvidia_x11}/lib/libcuda.so` is tied to the kernel module and will not work on the user's machine unless they happen to deploy exactly the same `linuxPackages` version

abysssol replied:
Thank you for letting me know about this. I created pr #323249, I hope you can review it and give further advice since I don't think I really know what I'm doing.
